### PR TITLE
Enum rollups

### PR DESCRIPTION
**Pull request by tilogaat**
What:
Laying down the infrastructure for rolling up enums

How: 
Added the buildRollupFromEnumRollups method in EnumRollup class. Plugged it in with the RollupRunnable class. 

TODO:
~~add test to see if serialization works for rollups too~~
~~improve ingestion testing for rollups~~

### CODE REVIEW

File: blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/service/RollupRunnableIntegrationTest.java [hunk 1]
```diff
@@ -95,7 +97,11 @@ public void setUp() throws Exception {
             BluefloodSetRollup rollup = new BluefloodSetRollup().withObject(i);
             metric = new PreaggregatedMetric(time, setLocator, ttl, rollup);
             preaggregatedMetrics.add(metric);
-            
+
```
Comment (VinnyQ) [inline blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/service/RollupRunnableIntegrationTest.java]: You should move the hashed value calculation here to a separate that can be repeated call for other tests that you're going to be writing ;)

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollup.java [hunk 2]
```diff
@@ -7,13 +10,22 @@
...
+
+        return this.withHashedEnumValue((long)valueName.hashCode(), value);
     }
 
     public BluefloodEnumRollup withHashedEnumValue(Long hashedEnumValue, Long value) {
-        this.hashedEnumValues2Count.put(hashedEnumValue, value);
+        long hashcount = 0;
+        if (this.hashedEnumValues2Count.containsKey(hashedEnumValue)) {
+            hashcount = this.hashedEnumValues2Count.get(hashedEnumValue);
+        }
```
Comment (VinnyQ) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollup.java]: I don't understand the line below, why are we putting value+hashcount in hashedEnumValues2Count?  Shoudn't it just be the hashcount?

Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollup.java]: @VinnyQ 

value is a bad name for this variable, it should be incomingCount or something to that effect. 
This test case explains why we need to add the existingCount with the incomingCount. 
https://github.com/rackerlabs/blueflood/pull/519/files#diff-437d2681df3fc60b82f4ed2954b87af2R37

In a single 5 min rollup period, if we see the enum value 'call_xyz_api:OK' first and again after say 2 mins 'call_xyz_api:OK', we want the map to store 'OK' => 2 and hash('OK') => 2. Is that right? If we do not add the 2 values, then the last OK will overwrite the first and the only value stored in the maps will be 'OK' => 1

Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollup.java]: I fixed variable names now and hopefully it is clearer.

File: blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/service/RollupRunnableIntegrationTest.java (hunk 1 above)
Comment (tilogaat) [inline blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/service/RollupRunnableIntegrationTest.java]: I just wanted to generate a long count for the values, so I just multiplied i+1 by 5 and cast it to long.

**Comment by VinnyQ** [comment]
:+1:  lgtm

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollup.java [hunk 3]
```diff
@@ -46,4 +58,24 @@ public boolean equals(Object obj) {
         return hashedEnumValues2Count.equals(other.hashedEnumValues2Count);
     }
 
+    public static BluefloodEnumRollup buildRollupFromEnumRollups(Points<BluefloodEnumRollup> input) throws IOException {
+        BluefloodEnumRollup enumRollup = new BluefloodEnumRollup();
+        for (Points.Point<BluefloodEnumRollup> point : input.getPoints().values()) {
+            BluefloodEnumRollup pointData = point.getData();
+            Map<Long, Long> incomingHashedEnums = pointData.getHashedEnumValuesWithCounts();
+            Map<Long, Long> currentHashedEnums = enumRollup.getHashedEnumValuesWithCounts();
```
Comment (GeorgeJahad) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollup.java]: It seems redundant to put this line in the for loop.  Wouldn't it work the same if currentHashedEnums was set once outside the loop? or am i missing something

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollup.java [hunk 4]
```diff
@@ -46,4 +58,24 @@ public boolean equals(Object obj) {
...
+
+            for (Long hash : incomingHashedEnums.keySet()) {
+                long incomingCount = incomingHashedEnums.get(hash);
+                if (currentHashedEnums.containsKey(hash)) {
+                    long currentCount = currentHashedEnums.get(hash);
+                    incomingCount+=currentCount;
+                }
+                enumRollup.getHashedEnumValuesWithCounts().put(hash, incomingCount);
```
Comment (GeorgeJahad) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollup.java]: since you've created "currentHashedEnums" as a synonym for "enumRollup.getHashedEnumValuesWithCounts()", couldn't you use it here as well?  In other words, instead of this line, wouldn't it be clearer to just do:
           currentHashedEnums .put(hash, incomingCount);

File: blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/service/RollupRunnableIntegrationTest.java [hunk 5]
```diff
@@ -187,6 +198,9 @@ private void testRolledupMetric(Locator locator, Class fullResClass, Class rollu
             } catch (InterruptedException e) {
             }
         }
+
+        reader.getDataToRoll(rollupClass, locator, range, CassandraModel.CF_METRICS_PREAGGREGATED_5M).getPoints();
```
Comment (GeorgeJahad) [inline blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/service/RollupRunnableIntegrationTest.java]: why is this line needed?

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollupTest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
+package com.rackspacecloud.blueflood.types;
```
Comment (GeorgeJahad) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollupTest.java]: File header?

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollupTest.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            else if (hash == 40L) {
+                Assert.assertEquals(40L, (long)hashCount);
+            }
+            else if (hash == 50L) {
+                Assert.assertEquals(10L, (long)hashCount);
+            }
+        }
+    }
+
+    private static <T> Points<T> asPoints(Class<T> type, long initialTime, long timeDelta, T... values) {
```
Comment (GeorgeJahad) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BluefloodEnumRollupTest.java]: You've added multiple copies of this method.   Can it be put in a utility class somewhere?

**Comment by GeorgeJahad** [comment]
+1
